src/interpreter/ByteCodeInterpreter.cpp

```diff
@@ -2698,156 +2710,183 @@ NEVER_INLINE void InterpreterSlowPath::setObjectPreComputedCaseOperationCacheMis
+            ASSERT(findResult.second->m_descriptor.isPlainDataProperty() && findResult.second->m_descriptor.isWritable());
+#endif
+
+            // set new cache item
+            newItem.m_cachedIndex = SetObjectInlineCacheData::CachedIndexMax;
+            newItem.m_cachedhiddenClassChainLength = cachedhiddenClassChain.size();
+            // +1 space for next object structure
+            newItem.m_cachedHiddenClassChainData = (ObjectStructure**)GC_MALLOC(sizeof(ObjectStructure*) * (newItem.m_cachedhiddenClassChainLength + 1));
+            memcpy(newItem.m_cachedHiddenClassChainData, cachedhiddenClassChain.data(), sizeof(ObjectStructure*) * newItem.m_cachedhiddenClassChainLength);
+            // caching the newly generated ObjectStructure at the end of m_cachedHiddenClassChainData
+            newItem.m_cachedHiddenClassChainData[newItem.m_cachedhiddenClassChainLength] = originalObject->structure();
+
+            block->m_inlineCacheDataSize += sizeof(size_t) * newItem.m_cachedhiddenClassChainLength;
+            currentCodeSizeTotal += sizeof(size_t) * newItem.m_cachedhiddenClassChainLength;
+
+            if (code->m_inlineCacheProtoTraverseMaxIndex == 0) {
+                // convert simple case to complex case
+                for (size_t i = 0; i < inlineCache->m_cache.size(); i++) {
+                    block->m_inlineCacheDataSize += sizeof(size_t);
+                    currentCodeSizeTotal += sizeof(size_t);
+
+                    // all previous cached data should be simple case
+                    ASSERT(inlineCache->m_cache[i].m_cachedhiddenClassChainLength == 1);
+                    auto oldClass = inlineCache->m_cache[i].m_cachedHiddenClass;
+                    inlineCache->m_cache[i].m_cachedHiddenClassChainData = (ObjectStructure**)GC_MALLOC(sizeof(ObjectStructure*));
+                    inlineCache->m_cache[i].m_cachedHiddenClassChainData[0] = oldClass;
+                }
             }
+            code->m_inlineCacheProtoTraverseMaxIndex = std::max(cachedhiddenClassChain.size() - 1, (size_t)code->m_inlineCacheProtoTraverseMaxIndex);
         }
-        code->m_inlineCacheProtoTraverseMaxIndex = std::max(cachedhiddenClassChain.size() - 1, (size_t)code->m_inlineCacheProtoTraverseMaxIndex);
-    }
 
-    return;
+        // finally, insert a valid new cache item at the end
+        // because an exception could occur ahead which makes insertion of cache item invalid
+        if (inlineCache->m_cache.size() > SetObjectInlineCacheData::MaxCacheCount) {
+            for (size_t i = inlineCache->m_cache.size() - 1; i > 0; i--) {
+                inlineCache->m_cache[i] = inlineCache->m_cache[i - 1];
+            }
+        } else {
+            inlineCache->m_cache.insert(0, SetObjectInlineCacheData());
+            block->m_inlineCacheDataSize += sizeof(SetObjectInlineCacheData);
+            currentCodeSizeTotal += sizeof(SetObjectInlineCacheData);
+        }
```

Comment:
Adding a new cache item at the end of `setObjectPreComputedCaseOperationSlowCase` operation because exceptions could arise in the middle of its execution, which may lead to insertion of an imperfact cache item.